e3sm_diags/driver/utils/dataset_xr.py

```diff
@@ -1202,13 +1144,8 @@ def _get_time_slice(self, ds: xr.Dataset, filename: str) -> slice:
         ValueError
             If invalid date range specified for test/reference time series data.
         """
-        start_yr_int = int(self.start_yr)
-        end_yr_int = int(self.end_yr)
-
-        # Get the available start and end years from the file name.
-        # Example: {var}_{start_yr}01_{end_yr}12.nc
-        var_start_year = int(filename.split("/")[-1].split("_")[-2][:4])
-        var_end_year = int(filename.split("/")[-1].split("_")[-1][:4])
+        start_yr_int, end_yr_int = int(self.start_yr), int(self.end_yr)
+        var_start_year, var_end_year = self._parse_years_from_filepaths(filepaths)
```

Comment:
@tomvothecoder can we parse_years based on ds_subset, so that we don't need this `_parse_years_from_filepaths` function?

Reply:
In this function, `ds_subset` has only been subsetted for specific variables and `.load()` is called. The time slice is determined using the filenames, then `ds_subset` is time sliced.

Reply:
To clarify, are you asking if we can parse the years directly from the time coordinates of the `xr.Dataset` object, since we've already opened up the files? I think that would be possible, which would eliminate the need to parse the years from the filepaths.

Reply:
I found that some datasets end on the first month of the next year, which means getting the year from the time coordinates will not work unless we check for this case and subtract 1 year.

Example:
```python
import xarray as xr

# Files
# --------------------
# Start year is 1985 and end year is 1988
filepaths = [
    "/lcrc/group/e3sm/ac.forsyth2/zppy_min_case_e3sm_diags_cdat_migrated_output/test-diags-no-cdat-20240917/v3.LR.historical_0051/post/rof/native/ts/monthly/2yr/RIVER_DISCHARGE_OVER_LAND_LIQ_198501_198612.nc",
    "/lcrc/group/e3sm/ac.forsyth2/zppy_min_case_e3sm_diags_cdat_migrated_output/test-diags-no-cdat-20240917/v3.LR.historical_0051/post/rof/native/ts/monthly/2yr/RIVER_DISCHARGE_OVER_LAND_LIQ_198701_198812.nc",
]

# Open the datasets
# --------------------
ds = xr.open_mfdataset(filepaths)
years = ds.time.dt.year

# 1985 is the first year
print(years[0].item())
# 1989 is the last year
print(years[-1].item())

```